Implemented combo boxes for game creation

blackshadowshade opened the pull request:
Fixes #108. Fixes #1316.

I'm submitting this pull request incomplete so that testers can have their say. I request that this be put up as a dev site to allow testing on various devices.

Before this pull request is accepted, I'd like:
- [x] ~~the QUnit stubs to be filled in~~
- [x] ~~@AdmiralJota to have his say on the customisation of the CSS~~
- [x] ~~functionality to be tested on various devices~~
- [x] ~~any potential conflicts with currently open pull requests to be resolved~~
- [x] ~~CircleCI to pass~~

Many of the changes in this pull request are a generic refactoring of Newgame.actionCreateGame, which was a monolith function with 242 lines of code.

I've added the external Chosen JS plugin (https://harvesthq.github.io/chosen/) which is by default disabled for iPods, iPhones, and Android devices signalling that they are mobile devices (https://github.com/harvesthq/chosen/pull/1388). We can artificially re-enable these devices, but I'm starting out with the default behaviour.

AdmiralJota commented:
Did you need me to review the javascript source changes, or should I wait for the dev site and just review how the CSS looks?

blackshadowshade commented:
I'd suggest waiting for the dev site and taking a look at it. I'm happy for you to take a look at the JS too, if you're motivated, but mostly, I value your judgement on look and feel.

blackshadowshade commented:
While testing, I noted that the button filters don't currently update the contents of the button combo box. I'll add that to my to-do list.

blackshadowshade commented:
Okay, filtering issue has been resolved. Button selectors now work correctly.

blackshadowshade commented:
QUnit tests added and pull request rebased.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/1451/ (if it passes)

blackshadowshade commented on src/ui/create_game.html:
```diff
@@ -28,7 +29,7 @@
         cache: true,
         success: function() {
           Loader.loadScripts([
-            'js/extern/jquery-ui.min.js',
```
Comment on removing the JQuery-UI library from here:

I think that we're not actually using the JQuery-UI library in the create_game.html page. I checked through the commit history, and the inclusion of this library dates all the way back to commit 1ed3976011024b6c652ddabe78a116093bb5e4bb, back in April 2013. It was only added to create_game.html to enable autocomplete. In other places, it's used for the date picker.

blackshadowshade commented:
@cgolubi1, I'm running into trouble with Jenkins. The last couple of runs have taken more than twice as long as previous runs (see http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/buildTimeTrend), and this current Jenkins run failed based on a timed-out waitFor() in the QUnit tests.

Is any of this related to the fixes for Meltdown/Spectre, or does Jenkins just need a restart?

cgolubi1 commented:
I rebooted jenkins - let's see if that helps anything.

blackshadowshade commented:
Thanks. Running Jenkins now.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/1452/ (if it passes)

blackshadowshade commented:
So, same problem, and still a Jenkins run of almost 3 hours duration.

blackshadowshade commented:
It's interesting to see where a lot of the time goes.

13 minutes on the PHP mess detector.
11 minutes on compiling the PHPDoc.
42 minutes on 'responderTest::test_request_invalid'
10 minutes on 'responderTest::test_request_createGame'
18 minutes on generating the PHP code coverage report

blackshadowshade commented:
Also, there is an odd failure:

ERROR: Directory '/var/lib/jenkins/jobs/buttonmen-blackshadowshade/workspace/build/code-browser' exists but failed copying to '/var/lib/jenkins/jobs/buttonmen-blackshadowshade/builds/2018-01-10_03-24-10/htmlreports/Code_Browser'.

cgolubi1 commented:
Okay, well, i tried stopping and starting it to get the instance to migrate, and it hasn't come back up yet.  I'm somewhat out of time for the morning, so sorry if i left things in a worse state, but i'll try to poke at it more tonight.

cgolubi1 commented:
Okay, it's back up now, and the shell seems a little zippier at least, so check out jenkins and let's see where we are.

blackshadowshade commented:
New Jenkins run: http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/1453/ (if it passes)

blackshadowshade commented:
Okay, it failed again, but at least we're back down to 44 minutes for a run. Time to go and have a look for what's causing this failure.

blackshadowshade commented:
I've added a no-op change and rebased. Hopefully this stops Jenkins doing any caching that was previously occurring.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/1454/ (if it passes)

blackshadowshade commented:
So, this continues to fail, but it passes fine on both Safari and Firefox on Mac.

I guess the next step is for someone to clone my branch, run Jenkins on it, and determine whether the failure occurs then too.

cgolubi1 commented:
Hmm, i'll try that when i get a chance, but IMO the fact that the failure is happening in the qunit tests with:
```
15:04:25      [exec] Running test: test_Newgame.getNewgameData
15:04:25      [exec] Running test: test_Newgame.showPage
15:06:51      [exec] 'waitFor()' timeout
15:06:51 
```

means you have a real bug in the test or the code.  Maybe something isn't finishing rendering or isn't getting cleaned up or something?  Do you see any JS warnings at all when you run the QUnit test locally using e.g. firebug?

blackshadowshade commented:
When I run the QUnit tests on Firefox Developer Edition, I get the following warnings:
- "Using //@ to indicate sourceMappingURL pragmas is deprecated. Use //# instead" (from jquery-1.9.0.min.js)
- "Use of getPreventDefault() is deprecated.  Use defaultPrevented instead." (from 
jquery.js:3308:28)
- "Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user’s experience. For more help http://xhr.spec.whatwg.org/" (from 
jquery.js:8441:6)

Other than those, no warnings. The third might *possibly* be relevant, I guess ...

blackshadowshade commented:
Oh, when I run that specify test (test_Newgame.showPage) on its own, there's no warning.

Also, when I run the previous test on its own, there's no warning either.

cgolubi1 commented:
Agreed.  QUnit tests pass on my browser.  So maybe something is different between what FF57 is doing and what phantom-js is doing.  I'll try to look into this more this weekend.

blackshadowshade commented:
Interesting that the QUnit tests also hang with CircleCI. We're going to need to find a way to address this problem.

blackshadowshade commented:
@cgolubi1, I had a thought about this. Is there somewhere where I've forgotten to specify a path to the new library for the testing rig?

blackshadowshade commented:
Okay, based on your feedback, I've added the path to the Chosen library to phantom-index.html, and then rebased. Let's see what CircleCI says about it now.

cgolubi1 commented:
shadowshade: we can take away the "incomplete" label, right?

Jota: yep, i need to stand up a dev site.  Starting that now, but i may not finish this morning.  I'll update if i do.

blackshadowshade commented:
Label 'incomplete' removed. :)

cgolubi1 commented:
Okay, the dev site is now actually up.  Irilyth, can you sync up momentary == 54.92.155.223 when you have a moment?

blackshadowshade commented:
Okay, a cursory look at the site shows that I can create a third-party game successfully on a MacBook Pro, using Safari 11.1.2.

irilyth commented:
> Okay, the dev site is now actually up. Irilyth, can you sync up momentary == 54.92.155.223 when you have a moment?

Done!

blackshadowshade commented:
So, an open call to all testers (especially @AdmiralJota): please test these changes on multiple devices and let me know what you think.

AdmiralJota commented:
Looks like the "Create another game?" button after you create a game doesn't have the combo boxes populated, at least in Firefox 52.8.1 in Windows 7.

irilyth commented:
Ubuntu Chrome Version 68.0.3440.84 (Official Build) (64-bit): The Create Game page looks good in general! One thing: On the game creation page, when I click on "Change first player", neither the Player 1 nor Player 2 box has this new UI.

Android Chrome Version 67.0.3396.87: Doesn't seem to have changed anything; unless I check the "Desktop site" menu item, and then it works as expected.

irilyth commented:
> Looks like the "Create another game?" button after you create a game doesn't have the combo boxes populated, at least in Firefox 52.8.1 in Windows 7.

Whoops, yeah, Ubuntu Chrome Version 68.0.3440.84 too. Screenshot attached.
![bm](https://user-images.githubusercontent.com/4923138/43692666-71585444-98f7-11e8-9c4a-634e558e9bb7.png)

blackshadowshade commented:
Okay, confirmed those two bugs.

blackshadowshade commented:
Okay, I think I've fixed these two bugs. If CircleCI likes the changes, we should be able to update the test site.

cgolubi1 commented:
Updated the dev site with shadowshade's fixes.  Check it out, testers.

irilyth commented:
The other bug seems fixed now, yay.

When you type in the box, it seems to only filter from the start of the name. Is that configurable? If I type "bla", it filters to blackshadowshade, but if I type "sha", it doesn't find anyone. If that's just the way it is, that's fine, but if there's an option to have it match anywhere in the name, I think that might be an improvement.

blackshadowshade commented:
I think the pathway to getting this pull request accepted is to get approval as-is. I'm happy to tweak various settings in a follow-up pull request, including activating Chosen on all phones/tablets, and doing autofill based on more lenient matching.

blackshadowshade commented:
From https://harvesthq.github.io/chosen/options.html, it appears that setting the 'search_contains' flag to TRUE is all we would need to do to get partial matching happening.

cgolubi1 commented:
Shadowshade: can you summarize the status here?  What bugs/concerns are outstanding, what should testers still be looking for?

blackshadowshade commented:
I believe that this should be ready to ship (after a rebase).

If you want me to add the option of filtering based on partial matching anywhere within a word, I can do that easily by setting an option in Chosen.

blackshadowshade commented:
Oh, there is one thing: we haven't had anyone explicitly say that they've tested this on an iPhone or iPad.

dwvanstone commented:
> Updated the dev site with shadowshade's fixes. Check it out, testers.

I tried 54.92.155.223 on an iPhone - it doesn't act any different from the live site in terms of selecting players or buttons. Is that correct?

irilyth commented:
DVS; If you have a way to turn on the "desktop version" of the site in the browser on your iPhone, it might be interesting to see what that does.

But I think the expected effect at this point is that mobile devices shouldn't see any difference from the regular site.

dwvanstone commented:
irilyth: yes, I can turn on the "desktop version" on my iPhone. That allows me the type-ahead feature!

cgolubi1 commented:
Okay, sorry i didn't get my act together to ask this question before the September release, but, hey, more incentive to have an October release: is this ready to go?  Anyone who's tested it know of a reason it's not ready to go?

blackshadowshade commented:
I was going to see how easy it was to set that 'search_contains' flag. I'll take a look at it in the next day or two.

cgolubi1 commented:
(It's not incomplete, but set the flag for my convenience because Shadowshade wants to do more work on this.)

blackshadowshade commented:
Okay, I've updated the .chosen() calls to add the 'search_contains' flag.

My limited testing shows that searches now do partial matching.

@cgolubi1, this should be ready for updating the dev site once CircleCI passes.

cgolubi1 commented:
Dev site updated.  Looks good to me at a glance; someone else want to verify?

dwvanstone commented:
It's looking good at a glance for me too (Firefox/Linux). A consequence I hadn't realized is that you can search for a particular die type such as "B(20)" and find all buttons with that die. (Or, technically, with that sequence of characters in the button description.) But that seems like a nice feature.

blackshadowshade commented:
(Oooh, I've just discovered task lists under Markup. Take a look at the first post, and also the display of the pull request on the list of pull requests!)

cgolubi1 commented:
Hey, cool, task lists.  Shadowshade, can you recap exactly what you want Jota to weigh in on?

blackshadowshade commented:
So, I can't remember exactly, but I think it was just a general callout to @AdmiralJota to take a look at the look and feel. I'm happy to not make that a blocker for accepting this pull request.  :)

irilyth commented:
Looks good to me, in Chrome on Linux and Android!

blackshadowshade commented:
Yes, of course, #2315 should be pretty easy once we get this pull request accepted.

In fact, if you wanted to do that yourself, it wouldn't be difficult.  ;)

cgolubi1 commented:
Okay, we're good to go here.  Go ahead and rebase this down to one commit.

blackshadowshade commented:
Rebased.
